### ISSUE TITLE
Investigate TypeScript 5 compatibility

### ISSUE THREAD

**Issue by simonhaenisch** [simonhaenisch/prettier-plugin-organize-imports#94]
https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#case-insensitive-import-sorting-in-editors

Lots of new options that potentially need to be exposed.

**Comment by maplesteve** [comment]
We have 40+ repos with failing prettier checks since TypeScript v5. Up until then everything was fine (prettier + 
prettier-plugin-organize-imports), so any "fix" to restore the former behaviour would be great.

**Comment by simonhaenisch** [comment]
Ok cool do you have any details with how they are failing exactly? Can they be auto-fixed? It might just be that it failed because the default import sort order changed?

**Comment by simonhaenisch** [comment]
Ok I see, I think it depends on whether you have VS Code set up to use the workspace's TypeScript version or the built-in one. See this guide for changing it: https://code.visualstudio.com/docs/typescript/typescript-compiling#_using-newer-typescript-versions.

Anyway, so you'd rather want an option here that allows you to go back to the old sorting behavior? I'll have to check first how these options need to be passed, not sure I'll have any time for it this week tbh.

**Comment by Zamiell** [comment]
> we do not want to do this, because a lot of repos are affected

Prettier `--write` should be able to fix everything automatically, even if a lot of repos are affected. Shouldn't we expect breaking changes like this when we upgrade our monorepos to TypeScript to 5.0? That does not strike me as a valid reason for requesting a new option from the plugin.

**Comment by maplesteve** [comment]
> Shouldn't we expect breaking changes like this when we upgrade our monorepos to TypeScript to 5.0?

TypeScript version numbers do not follow SemVer.

> That does not strike me as a valid reason for requesting a new option from the plugin.

I'm not requesting anything; just exploring the possibility to keep the former behaviour.

**Comment by Zamiell** [comment]
> TypeScript version numbers do not follow SemVer.

Indeed. But the consequence of that is you should NEVER update your TypeScript version unless you want breaking changes in your codebase.

I think the more salient point is that naturally, if one choses to update their TypeScript dep, they will generally expect new breaking type-related errors from the compiler. But they would not necessarily expect new breaking formatting changes from Prettier, since Prettier is an altogether different tool from TypeScript. So it is reasonable to be surprised when this turns out to be the case.

With that said, I would still argue that we should not be allowed to configure anything. Instead, we should expect formatting errors when we upgrade our TypeScript dep (even if it is counterintuitive), and if that is undesired, then we should not update our TypeScript dep.

**Comment by sven-meyer-wetter-com** [comment]
I've the same problem as @maplesteve but with IntelliJ. For me it's kinda weird. I can fix the issues running prettier --write, but i've also configured in IntelliJ that prettier runs as a save action. And when changing a file, prettier runs (and the plugin) and changes the  import order statements back to the previous order so that prettier checks are failing again. Example

Original file contents:
```javascript
export * from "./types";
export * from "./UrlParams";
export * from "./useUrlParams";
``` 
prettier --check complains

Contents after prettier --write
```javascript
export * from "./UrlParams";
export * from "./types";
export * from "./useUrlParams";
``` 
prettier --check passes

Now I add a space to the end of the file for example and trigger a reformat (and reorganize).

File contents:
```javascript
export * from "./types";
export * from "./UrlParams";
export * from "./useUrlParams";
``` 

Optimize imports in IntelliJ as a save action is turned off, only prettier runs. Typescript is used from projects node_modules. Any idea how this could happen?

**Comment by simonhaenisch** [comment]
Hey have you tried restarting IntelliJ? Maybe it was using a cached TS version from before you changed it to use the project TS?

**Comment by sven-meyer-wetter-com** [comment]
Yes, having this problem since the release because I tried to update to v5 on the first day. Now I'm still on 4.9.x

**Comment by DennisRosen** [comment]
I'm having similar problems using prettier in vs code with the prettier extension - after upgrading typescript this plugin stops working in vs code. Running prettier in the cmd line works fine. Both vs code's internal typescript and the locally installed typescript are set to 5.0.2.

**Comment by simonhaenisch** [comment]
Could you maybe search or open a discussion or issue in the Prettier VS Code/IntelliJ extension repo, to ask why the extension wouldn't run the TS version of the project?

**Comment by simonhaenisch** [comment]
Actually I just remembered this https://github.com/simonhaenisch/prettier-plugin-organize-imports/issues/84#issuecomment-1313669344:

> I have no idea why the Prettier extension runs prettier from the root directory and not the current VS Code project root

I feel like this could be causing this problem because it might not be able to resolve the project's `node_modules`, i.e. local TS version? Don't have time to investigate this currently though. Maybe try whether you can get it fixed with the `prettierPath` plugin config option.

**Comment by sven-meyer-wetter-com** [comment]
In my IntelliJ, typescript and prettier are used from the projects node_modules folder as per settings in the UI.

**Comment by simonhaenisch** [comment]
Ok, could you please try patching `node_modules/prettier-plugin-organize-imports/index.js`, there's a block

```js
if (process.env.DEBUG) {
	console.error(error);
}
```

and comment out the if-block so it always logs the error, and then check your IntelliJ plugin logs if you can see anything? Feel free to add extra `console.log`s, e.g. to make sure the plugin definitely is executed.

**Comment by DennisRosen** [comment]
I messed around with the prettier plugin settings in vs code and suddenly this plugin works again! It started working specifically after I unchecked the "Use Editor Config" setting. After checking it again it keeps on working though 🤷‍♂️. Maybe the extension just needed to be reset somehow?

Strangely I cannot get any debug logs from the organize imports plugin through the vs code Prettier output window though. The logs from prettier itself seem to indicate that the correct version is run (the one installed in the local project's `node_modules`).

**Comment by simonhaenisch** [comment]
I've been using the plugin in some projects that I upgraded to TS 5 recently, and it was working just fine. The line you linked is from dev dependencies (just used during plugin development/testing). Once you install the plugin into your project, it uses the `typescript` of your project (just has to match what's requested via peer dependencies).

So bumping the version in dev dependencies doesn't change anything except running the tests with TS 5 (but I'm pretty confident they would pass). Feel free to open a PR if you want to try ;) I still need to set up Dependabot for this repo

**Comment by Shinigami92** [comment]
> except running the tests with TS 5

my comment aimed explicitly for that
I know how `devDependencies` are working :wink:, I'm also an author of a plugin where I have a dev/peer-Dependency where tests are running with

**Comment by Shinigami92** [comment]
I tried to add a test case, but can not reproduce it and reported the bug upstream to TypeScript itself

https://github.com/microsoft/TypeScript/issues/54068

**Comment by ChristopheVandePoel** [comment]
Not sure if it is of any help, but we fixed it by pointing VSCode explicitly toward the installed prettier version:

```
// ./vscode/settings.json
{
  // some other vscode config stuff
  "prettier.prettierPath": "./node_modules/prettier"
}
```

Intellij kept working fine.

**Comment by simonhaenisch** [comment]
Closing this now as it seems to work fine with TypeScript 5 from what I can tell.
